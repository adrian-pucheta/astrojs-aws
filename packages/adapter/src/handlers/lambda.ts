import { Buffer } from "node:buffer"
import { URL } from "node:url"

import { polyfill } from "@astrojs/webapi"
import { App } from "astro/app"
import { APIGatewayProxyHandlerV2, APIGatewayProxyResultV2 } from "aws-lambda"

import { parseContentType } from "../helpers"

polyfill(globalThis, {
  exclude: "window document",
})

export default function handler(
  app: App,
  knownBinaryMediaTypes: Set<string>,
): APIGatewayProxyHandlerV2 {
  return async function (event): Promise<APIGatewayProxyResultV2> {
    const {
      body,
      headers,
      rawPath,
      rawQueryString,
      requestContext,
      isBase64Encoded,
    } = event

    // convert aws apigateway event to node request
    const scheme = headers["x-forwarded-protocol"] || "https"
    const host = headers["x-forwarded-host"] || headers.host
    const qs = rawQueryString.length > 0 ? `?${rawQueryString}` : ""
    const url = new URL(`${rawPath}${qs}`, `${scheme}://${host}`)
    const encoding = isBase64Encoded ? "base64" : "utf8"
    const request = new Request(url.toString(), {
      method: requestContext.http.method,
      headers: new Headers(headers as HeadersInit),
      body: typeof body === "string" ? Buffer.from(body, encoding) : body,
    })

    const routeData = app.match(request, { matchNotFound: true })
    if (!routeData) {
      return {
        statusCode: 404,
        body: "Not found",
      }
    }

    // astro render
    const rendered = await app.render(request, routeData)

    // convert node response to apigateway response
    const contentType = parseContentType(rendered.headers.get("content-type"))
    const responseIsBase64Encoded = knownBinaryMediaTypes.has(contentType)
    return {
      statusCode: rendered.status,
      headers: Object.fromEntries(rendered.headers.entries()),
      cookies: Array.from(app.setCookieHeaders(rendered)),
      body: responseIsBase64Encoded
        ? Buffer.from(await rendered.arrayBuffer()).toString("base64")
        : await rendered.text(),
      isBase64Encoded: responseIsBase64Encoded,
    }
  }
}
