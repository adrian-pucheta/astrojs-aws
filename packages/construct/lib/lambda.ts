import { HttpApi, HttpMethod } from "@aws-cdk/aws-apigatewayv2-alpha"
import { HttpUrlIntegration } from "@aws-cdk/aws-apigatewayv2-integrations-alpha"
import {
  BehaviorOptions,
  CachePolicy,
  Distribution,
} from "aws-cdk-lib/aws-cloudfront"
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs"
import { Bucket } from "aws-cdk-lib/aws-s3"
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment"
import { Construct } from "constructs"

import { AstroSiteConstruct } from "./construct"
import { LambdaAstroSiteProps } from "./types"

export class LambdaAstroSite extends AstroSiteConstruct {
  private readonly bucket: Bucket
  private readonly function: NodejsFunction
  private readonly httpApi: HttpApi
  private readonly distribution?: Distribution
  readonly bucketArn: string
  readonly bucketName: string
  readonly functionArn: string
  readonly functionName: string
  readonly distributionId: string
  readonly domains: string[]

  constructor(scope: Construct, id: string, props: LambdaAstroSiteProps) {
    super(scope, id)

    const cfEnabled = props.cfOptions != undefined
    const domains = []

    this.bucket = this.newBucket(this, !cfEnabled)
    this.function = this.newFunction(
      this,
      props.serverEntry,
      props.serverOptions,
    )

    this.httpApi = this.newHttpApiGw(this, this.function, props.gwOptions)

    if (!cfEnabled) {
      domains.push(`http://${this.bucket.bucketWebsiteDomainName}`)

      // add http integration routes
      const routes = this.parseRoutesFromDir(props.staticDir)
      for (const [path, route] of Object.entries(routes)) {
        this.httpApi.addRoutes({
          path: path,
          methods: [HttpMethod.GET],
          integration: new HttpUrlIntegration(
            path,
            `http://${this.bucket.bucketWebsiteDomainName}${route}`,
          ),
        })
      }
    } else {
      const defaultBehavior: BehaviorOptions = {
        origin: this.newHttpApiGatewayOrigin(this.httpApi),
        cachePolicy: CachePolicy.CACHING_DISABLED,
      }
      this.distribution = this.newDistribution(
        this,
        defaultBehavior,
        props.cfOptions,
      )

      const routes = this.parseRoutesFromDir(props.staticDir, true)
      for (const [path] of Object.entries(routes)) {
        this.distribution.addBehavior(path, this.newS3Origin(this, this.bucket))
      }

      domains.push(`https://${this.distribution.distributionDomainName}`)
      domains.push(`https://${props.cfOptions?.domain ?? ""}`)
    }

    new BucketDeployment(this, "BucketDeployment", {
      sources: [Source.asset(props.staticDir)],
      destinationBucket: this.bucket,
      distribution: this.distribution,
    })

    this.bucketArn = this.bucket.bucketArn
    this.bucketName = this.bucket.bucketName
    this.functionArn = this.function.functionArn
    this.functionName = this.function.functionName
    this.distributionId = this.distribution?.distributionId ?? ""
    this.domains = domains
  }
}
