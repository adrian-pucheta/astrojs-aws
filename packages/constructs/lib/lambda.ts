import {
  AddRoutesOptions,
  HttpApi,
  HttpMethod,
  HttpRoute,
  HttpStage,
  HttpStageOptions,
} from "@aws-cdk/aws-apigatewayv2-alpha"
import { HttpLambdaIntegration } from "@aws-cdk/aws-apigatewayv2-integrations-alpha"
import { CfnOutput, Fn } from "aws-cdk-lib"
import {
  AddBehaviorOptions,
  AllowedMethods,
  Distribution,
  IOrigin,
  ViewerProtocolPolicy,
} from "aws-cdk-lib/aws-cloudfront"
import { HttpOrigin, S3Origin } from "aws-cdk-lib/aws-cloudfront-origins"
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs"
import { Bucket, LifecycleRule } from "aws-cdk-lib/aws-s3"
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment"
import { Construct } from "constructs"

import { AstroSiteConstruct } from "./construct"
import { LambdaAstroSiteProps } from "./types"

export class LambdaAstroSite extends AstroSiteConstruct {
  /**
   * S3 bucket
   */
  readonly bucket: Bucket
  /**
   * Lambda function handler
   */
  readonly handler: NodejsFunction
  /**
   * HTTP API
   */
  readonly httpApi: HttpApi
  /**
   * CloudFront distribution
   */
  readonly distribution: Distribution

  constructor(scope: Construct, id: string, props: LambdaAstroSiteProps) {
    super(scope, id)

    this.bucket = new Bucket(this, "LambdaAstroSiteBucket", {
      ...props.bucketOptions,
    })
    new BucketDeployment(this, "LambdaAstroSiteBucketDeployment", {
      ...props.bucketDeploymentOptions,
      sources: [Source.asset(props.staticDir)],
      destinationBucket: this.bucket,
    })

    const runtime = this.strToRuntime(props.serverOptions?.runtime)
    this.handler = new NodejsFunction(this, "LambdaAstroSiteHandler", {
      ...props.serverOptions,
      entry: props.serverEntry,
      runtime: runtime,
    })

    this.httpApi = new HttpApi(this, "LambdaAstroSiteHttpApi", {
      ...props.httpApiOptions,
    })
    const integration = new HttpLambdaIntegration(
      "HttpLambdaIntegration",
      this.handler,
    )
    this.httpApi.addRoutes({
      path: "/",
      methods: [HttpMethod.ANY],
      integration: integration,
    })

    this.distribution = new Distribution(this, "LambdaAstroSiteDistribution", {
      ...props.distributionOptions,
      defaultBehavior: {
        origin: new HttpOrigin(
          Fn.select(1, Fn.split("://", this.httpApi.apiEndpoint ?? "")),
        ),
        allowedMethods: AllowedMethods.ALLOW_ALL,
        viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      },
    })
    const routes = this.parseRoutesFromDir(props.staticDir)
    for (const route of routes) {
      this.distribution.addBehavior(route, new S3Origin(this.bucket), {
        viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      })
    }

    new CfnOutput(this, "CloudFrontDomainName", {
      value: this.distribution.domainName,
    })
  }
  /**
   * Add a lifecycle rule to the bucket
   *
   * @param rule
   */
  addBucketLifecycleRule(rule: LifecycleRule) {
    this.bucket.addLifecycleRule(rule)
  }
  /**
   * Adds a new behavior to this distribution for the given pathPattern.
   *
   * @param pathPattern the path pattern (e.g., 'images/*') that specifies which requests to apply the behavior to.
   * @param origin the origin to use for this behavior
   * @param behaviorOptions the options for the behavior at this path.
   */
  addDistributionBehavior(
    pathPattern: string,
    origin: IOrigin,
    behaviorOptions?: AddBehaviorOptions,
  ) {
    this.distribution?.addBehavior(pathPattern, origin, behaviorOptions)
  }
  /**
   * Add a new stage.
   */
  addHttpApiStage(id: string, options: HttpStageOptions): HttpStage {
    return this.httpApi.addStage(id, options)
  }
  /**
   * Add multiple routes that uses the same configuration. The routes all go to the same path, but for different
   * methods.
   */
  addHttpApiRoutes(options: AddRoutesOptions): HttpRoute[] {
    return this.httpApi.addRoutes(options)
  }
}
