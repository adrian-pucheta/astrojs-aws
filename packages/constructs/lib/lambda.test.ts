import path from "path"
import { fileURLToPath } from "url"

import { App, Stack } from "aws-cdk-lib"
import { Template } from "aws-cdk-lib/assertions"
import { describe, expect, test } from "vitest"

import { LambdaAstroSite } from "./lambda"

const __dirname = path.dirname(fileURLToPath(import.meta.url))

describe("Test LambdaAstroSite", () => {
  test("Expect match snapshot", () => {
    const stack = new Stack(new App(), "TestStack")
    new LambdaAstroSite(stack, "TestConstruct", {
      entry: path.join(__dirname, "../tests/testdata/fake.ts"),
    })
    const template = Template.fromStack(stack).toJSON()
    expect(template).toMatchSnapshot()
    expect(1).toBe(1)
  })
})
