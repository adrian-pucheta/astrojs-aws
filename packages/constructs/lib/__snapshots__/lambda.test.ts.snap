// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test LambdaAstroSite > Expect match snapshot 1`] = `
{
  "Outputs": {
    "TestConstructAPIEndpoint8D83142B": {
      "Value": {
        "Fn::GetAtt": [
          "TestConstructHttpApiDD05EA19",
          "ApiEndpoint",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestConstructHandlerFFEC74AC": {
      "DependsOn": [
        "TestConstructHandlerServiceRoleC1B4872E",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b4349d2e7ca08ef1b005e6e30b06256ff57aeab10edeb70ad5b5be5cf9ad45d1.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestConstructHandlerServiceRoleC1B4872E",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestConstructHandlerServiceRoleC1B4872E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestConstructHttpApiANY7337DC9E": {
      "Properties": {
        "ApiId": {
          "Ref": "TestConstructHttpApiDD05EA19",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "ANY /",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "TestConstructHttpApiANYAstroIntegration557C532A",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "TestConstructHttpApiANYAstroIntegration557C532A": {
      "Properties": {
        "ApiId": {
          "Ref": "TestConstructHttpApiDD05EA19",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "TestConstructHandlerFFEC74AC",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "TestConstructHttpApiANYAstroIntegrationPermissionCF7F49A3": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "TestConstructHandlerFFEC74AC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "TestConstructHttpApiDD05EA19",
              },
              "/*/*/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestConstructHttpApiDD05EA19": {
      "Properties": {
        "CorsConfiguration": {
          "AllowMethods": [
            "*",
          ],
          "AllowOrigins": [
            "*",
          ],
        },
        "Name": "HttpApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "TestConstructHttpApiDefaultStage63A0347E": {
      "Properties": {
        "ApiId": {
          "Ref": "TestConstructHttpApiDD05EA19",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
